File: mobile/src/services/api/world.ts
import { apiClient } from "./client";
import type { SubjectId, ChildProgress } from "../../../shared/types";

export interface SubjectProgressSummary {
  subject_id: SubjectId;
  level: number;
  accuracy_rate: number;
  questions_answered: number;
  last_played: string | null;
}

export interface WorldGreetingRequest {
  child_id: string;
  child_name: string;
  child_age: number;
  time_of_day: "morning" | "afternoon" | "evening";
  subject_progress: SubjectProgressSummary[];
}

export interface WorldGreetingResponse {
  greeting: string;
  recommended_zone: SubjectId;
  tts_url: string | null;
}

export const fetchWorldGreeting = async (
  req: WorldGreetingRequest,
): Promise<WorldGreetingResponse> => {
  const { data } = await apiClient.post<WorldGreetingResponse>(
    "/api/v1/world/greeting",
    req,
  );
  return data;
};

/** Convert ChildProgress (from game API) to the summary shape the world API expects. */
export function progressToSummary(p: ChildProgress): SubjectProgressSummary {
  return {
    subject_id: p.subject_id as SubjectId,
    level: p.level,
    accuracy_rate: p.accuracy_rate,
    questions_answered: p.questions_answered,
    last_played: p.last_played ?? null,
  };
}

File: backend/app/models/world.py
"""
World map models — session greeting and zone state.
"""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class SubjectProgressSummary(BaseModel):
    subject_id: str          # GameSubjectId value
    level: int = 1
    accuracy_rate: float = 0.0
    questions_answered: int = 0
    last_played: Optional[str] = None   # ISO 8601 string or None


class WorldGreetingRequest(BaseModel):
    child_id: str
    child_name: str
    child_age: int
    time_of_day: str                              # "morning" | "afternoon" | "evening"
    subject_progress: list[SubjectProgressSummary] = []


class WorldGreetingResponse(BaseModel):
    greeting: str             # Zoey's spoken greeting
    recommended_zone: str     # GameSubjectId value
    tts_url: Optional[str] = None
